src/Features/Core/Completion/AbstractCompletionService.cs

```diff
@@ -130,65 +135,146 @@ private class ProviderGroup
+        {
+            var displayNameToItemsMap = new Dictionary<string, List<CompletionItem>>();
+            CompletionItem builder = null;
+
+            foreach (var completionList in completionLists)
+            {
+                if (completionList != null)
+                {
+                    foreach (var item in completionList.Items.WhereNotNull())
+                    {
+                        // New items that match an existing item will replace it.  
+                        ReplaceExistingItem(item, displayNameToItemsMap, completionRules);
+                    }
+
+                    builder = builder ?? completionList.Builder;
+                }
+            }
+
+            if (displayNameToItemsMap.Count == 0)
+            {
+                return null;
+            }
+
+            var totalItems = displayNameToItemsMap.Values.Flatten().ToList();
+            totalItems.Sort();
+
+            // TODO(DustinCa): This is lossy -- we lose the IsExclusive field. Fix that.
+
+            return new CompletionList(totalItems, builder);
         }
 
-        private Dictionary<ICompletionProvider, int> GetCompletionProviderToIndex(IEnumerable<ICompletionProvider> completionProviders)
+        private static void ReplaceExistingItem(
+            CompletionItem item,
+            Dictionary<string, List<CompletionItem>> displayNameToItemsMap,
+            ICompletionRules completionRules)
         {
-            var result = new Dictionary<ICompletionProvider, int>();
+            // See if we have an item with 
+            var sameNamedItems = displayNameToItemsMap.GetOrAdd(item.DisplayText, s_createList);
+            for (int i = 0; i < sameNamedItems.Count; i++)
+            {
+                var existingItem = sameNamedItems[i];
+
+                Contract.Assert(item.DisplayText == existingItem.DisplayText);
+
+                if (completionRules.ItemsMatch(item, existingItem).Value)
+                {
+                    sameNamedItems[i] = Disambiguate(item, existingItem);
+                    return;
+                }
+            }
+
+            sameNamedItems.Add(item);
+        }
+
+        private static CompletionItem Disambiguate(CompletionItem item, CompletionItem existingItem)
+        {
+            // We've constructed the export order of completion providers so 
+            // that snippets are exported after everything else. That way,
+            // when we choose a single item per display text, snippet 
+            // glyphs appear by snippets. This breaks preselection of items
+            // whose display text is also a snippet (workitem 852578),
+            // the snippet item doesn't have its preselect bit set.
+            // We'll special case this by not preferring later items
+            // if they are snippets and the other candidate is preselected.
+            if (existingItem.Preselect && item.CompletionProvider is ISnippetCompletionProvider)
```

Comment:
Oh I know. That's a copy-and-paste on my part. I'm deferring thinking about how Snippets fit in for "future Dustin".